From f57f343af5279957f8df179a4c7a167bba635260 Mon Sep 17 00:00:00 2001
From: Igal Zeltser <izeltser@habana.ai>
Date: Thu, 08 Feb 2024 12:33:08 +0200
Subject: [PATCH] [DO NOT MERGE] create ppboot SA elf

add start and exception vectors in the beginning of the binary, so ARC
can jump and run it from reset vector.

mostly usefull for simulator

Signed-off-by: Igal Zeltser <izeltser@habana.ai>
Change-Id: Id8e8f2ea427adc80c12f687f9251303cc2b95fd5
---

diff --git a/app/ppboot/crt0_ppboot.S b/app/ppboot/crt0_ppboot.S
index d9d7025..0ffd55c 100644
--- a/app/ppboot/crt0_ppboot.S
+++ b/app/ppboot/crt0_ppboot.S
@@ -5,6 +5,58 @@
  * SPDX-License-Identifier:     GPL-2.0+
  */
 
+.global _exit_halt
+.type _exit_halt, @function
+.align 4
+_exit_halt:
+	; r0 contains exit code
+	flag 0x01
+	nop
+	nop
+	nop
+	j @_exit_halt
+	nop
+
+
+.section .ivt, "a", @progbits
+
+; Helper macro to define weak symbols to include into interrupt vector table.
+; User code may define those functions in them, so user function will be
+; referenced in the IVT. By default all handlers point to _exit_halt - so they
+; always cause application halt, because if application causes an exception or
+; interrupt, but doesn't set a handler for it - something is wrong in
+; application. Exception is "start" entry of IVT, which points to __start
+; function.
+#define IVT_ENTRY(name) \
+    .word name `\
+    .weak name `\
+    .set name, _exit_halt
+
+; handler's name,		number,	name,			offset in IVT (hex/dec)
+.word _start			; 0	program entry point	0x0		0
+IVT_ENTRY(memory_error)		; 1	memory_error		0x4		4
+IVT_ENTRY(instruction_error)	; 2	instruction_error	0x8		8
+IVT_ENTRY(EV_MachineCheck)	; 3	EV_MachineCheck		0xC		12
+IVT_ENTRY(EV_TLBMissI)		; 4	EV_TLBMissI		0x10		16
+IVT_ENTRY(EV_TLBMissD)		; 5	EV_TLBMissD		0x14		20
+IVT_ENTRY(EV_ProtV)		; 6	EV_ProtV		0x18		24
+IVT_ENTRY(EV_PrivilegeV)	; 7	EV_PrivilegeV		0x1C		28
+IVT_ENTRY(EV_SWI)		; 8	EV_SWI			0x20		32
+IVT_ENTRY(EV_Trap)		; 9	EV_Trap			0x24		36
+IVT_ENTRY(EV_Extension)		; 10	EV_Extension		0x28		40
+IVT_ENTRY(EV_DivZero)		; 11	EV_DivZero		0x2C		44
+IVT_ENTRY(EV_DCError)		; 12	EV_DCError		0x30		48
+IVT_ENTRY(EV_Maligned)		; 13	EV_Maligned		0x34		52
+IVT_ENTRY(EV_Ex14)		; 14	unused			0x38		56
+IVT_ENTRY(EV_Ex15)		; 15	unused			0x3C		60
+IVT_ENTRY(IRQ_Timer0)		; 16	Timer 0			0x40		64
+IVT_ENTRY(IRQ_Timer1)		; 17	Timer 1			0x44		68
+IVT_ENTRY(IRQ_18)		; 18				0x48		72
+IVT_ENTRY(IRQ_19)		; 19				0x4C		76
+IVT_ENTRY(IRQ_20)		; 20				0x50		80
+
+.section .text.__startup, "ax", @progbits
+
 /*
  * This is preboot entry point
  * Setting sp/fp/lr and jumping to C code
diff --git a/arch/fs1/arc/ppboot.lds.S b/arch/fs1/arc/ppboot.lds.S
index 1906acc..e407f4b 100644
--- a/arch/fs1/arc/ppboot.lds.S
+++ b/arch/fs1/arc/ppboot.lds.S
@@ -26,6 +26,10 @@
 ENTRY(_start)
 SECTIONS
 {
+	.ivt : {
+		KEEP (*(.ivt));
+	} > .sram
+
 	.text : {
 		. = ALIGN(8);
 		app/ppboot/crt0_ppboot.o (.text*)
